test/space/cu_ntuples_test.py

```diff
@@ -18,70 +18,151 @@
-        return np.random.randn(fn.size).astype(fn.dtype)
+        return np.random.uniform(-1, 1, fn.size).astype(fn.dtype)
     elif np.issubdtype(fn.dtype, np.integer):
         return np.random.randint(0, 10, fn.size).astype(fn.dtype)
     else:
-        return (np.random.randn(fn.size) +
-                1j * np.random.randn(fn.size)).astype(fn.dtype)
+        return (np.random.uniform(-1, 1, fn.size) +
+                1j * np.random.uniform(-1, 1, fn.size)).astype(fn.dtype)
 
-def _element(fn):
-    return fn.element(_array(fn))
 
 def _vectors(fn, n=1):
+    """Create a list of arrays and vectors in `fn`.
+
+    First arrays, then vectors.
+    """
     arrs = [_array(fn) for _ in range(n)]
 
     # Make Fn vectors
     vecs = [fn.element(arr) for arr in arrs]
     return arrs + vecs
 
+
+def _pos_array(fn):
+    """Create an array with positive real entries as weight in `fn`."""
+    return np.abs(_array(fn)) + 0.1
+
+
+def _sparse_matrix(fn):
+    """Create a sparse positive definite Hermitian matrix for `fn`."""
+    nnz = np.random.randint(0, int(ceil(fn.size**2/2)))
+    coo_r = np.random.randint(0, fn.size, size=nnz)
+    coo_c = np.random.randint(0, fn.size, size=nnz)
+    values = np.random.rand(nnz).astype(fn.dtype)
+    if fn.field == ComplexNumbers():
+        values += 1j * np.random.rand(nnz).astype(fn.dtype)
+    mat = sp.sparse.coo_matrix((values, (coo_r, coo_c)),
+                               shape=(fn.size, fn.size),
+                               dtype=fn.dtype)
+    # Make symmetric and positive definite
+    return mat + mat.conj().T + fn.size * sp.sparse.eye(fn.size,
+                                                        dtype=fn.dtype)
+
+
+def _dense_matrix(fn):
+    """Create a dense positive definite Hermitian matrix for `fn`."""
+    mat = np.asarray(np.random.rand(fn.size, fn.size), dtype=fn.dtype)
+    # Make symmetric and positive definite
+    return mat + mat.conj().T + fn.size * np.eye(fn.size, dtype=fn.dtype)
+
+
+def _element(fn):
+    return fn.element(_array(fn))
+
+
 @pytest.fixture(scope="module",
                 ids=['CudaRn float32'],
                 params=[odl.CudaRn(100)])
 def fn(request):
     return request.param
 
-@pytest.mark.skipif("np.float32 not in odl.CUDA_DTYPES")
-def test_init_cudantuples_f32():
-    #verify that the code runs
-    r3 = odl.CudaNtuples(3, dtype='float32')
-    r3.element()
+
+@pytest.fixture(scope="module",
+                ids=[' p = 2 ', ' p = 1 ', ' p = inf ', ' p = 0.5 ',
+                     ' p = 1.5 ', ' p = 3 '],
+                params=[2.0, 1.0, float('inf'), 0.5, 1.5, 3.0])
+def exponent(request):
+    return request.param
+
+
+# --- CUDA space tests --- #
+
+
+def test_init_cudantuples_dtypes():
+    # verify that the code runs
+    for dtype in odl.CUDA_DTYPES:
+        odl.CudaNtuples(3, dtype=dtype).element()
+        odl.CudaFn(3, dtype=dtype).element()
+
+
+def test_init_exponent(exponent):
+    for dtype in odl.CUDA_DTYPES:
```

Comment:
Couldnt be made in the context of a unified `FnBase` test since this tests the initialization of these `Fn`'s, but sure, that branch may be a good place to clean all of this up.